File: apps/backend/routes/services/api_result.py
# apps/backend/routes/services/api_result.py
from __future__ import annotations

from typing import Any, Dict, Optional, TypedDict, Union


class ApiOk(TypedDict):
    ok: bool
    data: Any


class ApiErr(TypedDict, total=False):
    ok: bool
    error: str
    details: Any


ApiResult = Union[ApiOk, ApiErr]


def ok(data: Any) -> ApiOk:
    return {"ok": True, "data": data}


def err(error: str, details: Optional[Any] = None) -> ApiErr:
    out: ApiErr = {"ok": False, "error": error}
    if details is not None:
        out["details"] = details
    return out


def coerce_error_message(e: Exception) -> str:
    return str(e) or e.__class__.__name__

File: apps/backend/shared/apiResult.ts
// apps/backend/shared/apiResult.ts
// Canonical API envelope used by ALL backend routes.

export type ApiOk<T> = {
  ok: true;
  data: T;
};

export type ApiErr = {
  ok: false;
  error: string;
  details?: any;
};

export type ApiResult<T> = ApiOk<T> | ApiErr;

export function ok<T>(data: T): ApiOk<T> {
  return { ok: true, data };
}

export function err(message: string, details?: any): ApiErr {
  return { ok: false, error: message, details };
}

// Optional: helper for unexpected exceptions
export function fromException(message: string, e: unknown): ApiErr {
  const details =
    e instanceof Error
      ? { name: e.name, message: e.message, stack: e.stack }
      : { value: e };
  return err(message, details);
}
